arangod/Cache/TransactionalCache.cpp

```diff
@@ -66,139 +67,150 @@ Finding TransactionalCache::find(void const* key, uint32_t keySize) {
     result.reportError(status);
   }
   recordStat(result.found() ? Stat::findHit : Stat::findMiss);
-  bucket->unlock();
 
   return result;
 }
 
 Result TransactionalCache::insert(CachedValue* value) {
   TRI_ASSERT(value != nullptr);
-  uint32_t hash = hashKey(value->key(), value->keySize());
+  bool maybeMigrate = false;
+  std::uint32_t hash = hashKey(value->key(), value->keySize());
 
   Result status;
-  TransactionalBucket* bucket;
   Table* source;
-  std::tie(status, bucket, source) = getBucket(hash, Cache::triesFast);
-  if (status.fail()) {
-    return status;
-  }
-
-  bool maybeMigrate = false;
-  bool allowed = !bucket->isBlacklisted(hash);
-  if (allowed) {
-    int64_t change = static_cast<int64_t>(value->size());
-    CachedValue* candidate = bucket->find(hash, value->key(), value->keySize());
-
-    if (candidate == nullptr && bucket->isFull()) {
-      candidate = bucket->evictionCandidate();
-      if (candidate == nullptr) {
-        allowed = false;
-        status.reset(TRI_ERROR_ARANGO_BUSY);
-      }
+  {
+    Table::BucketLocker guard;
+    std::tie(status, guard) = getBucket(hash, Cache::triesFast, false);
+    if (status.fail()) {
+      return status;
     }
 
+    TransactionalBucket& bucket = guard.bucket<TransactionalBucket>();
+    source = guard.source();
+    bool allowed = !bucket.isBlacklisted(hash);
     if (allowed) {
-      if (candidate != nullptr) {
-        change -= static_cast<int64_t>(candidate->size());
+      std::int64_t change = static_cast<std::int64_t>(value->size());
+      CachedValue* candidate = bucket.find(hash, value->key(), value->keySize());
+
+      if (candidate == nullptr && bucket.isFull()) {
+        candidate = bucket.evictionCandidate();
+        if (candidate == nullptr) {
+          allowed = false;
+          status.reset(TRI_ERROR_ARANGO_BUSY);
+        }
       }
 
-      _metadata.lockRead();
-      allowed = _metadata.adjustUsageIfAllowed(change);
-      _metadata.unlockRead();
-
       if (allowed) {
-        bool eviction = false;
         if (candidate != nullptr) {
-          bucket->evict(candidate, true);
-          if (!candidate->sameKey(value->key(), value->keySize())) {
-            eviction = true;
-          }
-          freeValue(candidate);
+          change -= static_cast<std::int64_t>(candidate->size());
+        }
+
+        {
+          SpinLocker metaGuard(SpinLocker::Mode::Read, _metadata.lock());
+          allowed = _metadata.adjustUsageIfAllowed(change);
```

Comment:
This has been a read-lock before as well, but wondering why a call to `adjustUsageIfAllowed` only requires a read-lock and not a write-lock.... Obviously `adjust...` modifies the state, so should it be using the write-lock instead?

Reply:
So we special-cased this to allow for simply read-locking the record. Internally it handles all the logic appropriately with a compare and swap on an atomic. Other changes require write-locking in order to get the logic correct, but the usage can be adjusted with a simple read lock and this extra synchronization.